File: library/src/main/java/com/mikepenz/aboutlibraries/entity/Library.java
package com.mikepenz.aboutlibraries.entity;

public class Library implements Comparable<Library> {

    private String definedName = "";

    private boolean internal = false;

    private String author = "";
    private String authorWebsite = "";
    private String libraryName = "";
    private String libraryDescription = "";
    private String libraryVersion = "";
    private String libraryWebsite = "";
    private License license;

    private boolean isOpenSource = true;
    private String repositoryLink = "";

    private String classPath = "";

    public Library() {

    }

    public Library(String author, String libraryName, String libraryDescription) {
        this.author = author;
        this.libraryName = libraryName;
        this.libraryDescription = libraryDescription;
    }

    public Library(String author, String libraryName, String libraryDescription, String libraryVersion) {
        this.author = author;
        this.libraryName = libraryName;
        this.libraryDescription = libraryDescription;
        this.libraryVersion = libraryVersion;
    }

    public String getDefinedName() {
        return definedName;
    }

    public void setDefinedName(String definedName) {
        this.definedName = definedName;
    }

    public String getAuthor() {
        return author;
    }

    public void setAuthor(String author) {
        this.author = author;
    }

    public String getAuthorWebsite() {
        return authorWebsite;
    }

    public void setAuthorWebsite(String authorWebsite) {
        this.authorWebsite = authorWebsite;
    }

    public String getLibraryName() {
        return libraryName;
    }

    public void setLibraryName(String libraryName) {
        this.libraryName = libraryName;
    }

    public String getLibraryDescription() {
        return libraryDescription;
    }

    public void setLibraryDescription(String libraryDescription) {
        this.libraryDescription = libraryDescription;
    }

    public String getLibraryVersion() {
        return libraryVersion;
    }

    public void setLibraryVersion(String libraryVersion) {
        this.libraryVersion = libraryVersion;
    }

    public String getLibraryWebsite() {
        return libraryWebsite;
    }

    public void setLibraryWebsite(String libraryWebsite) {
        this.libraryWebsite = libraryWebsite;
    }

    public License getLicense() {
        return license;
    }

    public void setLicense(License license) {
        this.license = license;
    }

    public boolean isOpenSource() {
        return isOpenSource;
    }

    public void setOpenSource(boolean isOpenSource) {
        this.isOpenSource = isOpenSource;
    }

    public String getRepositoryLink() {
        return repositoryLink;
    }

    public void setRepositoryLink(String repositoryLink) {
        this.repositoryLink = repositoryLink;
    }

    public boolean isInternal() {
        return internal;
    }

    public void setInternal(boolean internal) {
        this.internal = internal;
    }

    public String getClassPath() {
        return classPath;
    }

    public void setClassPath(String classPath) {
        this.classPath = classPath;
    }

    @Override
    public int compareTo(Library another) {
        return getLibraryName().compareToIgnoreCase(another.getLibraryName());
    }
}

File: library-core/src/main/java/com/mikepenz/aboutlibraries/entity/Library.kt
package com.mikepenz.aboutlibraries.entity


data class Library(
        var definedName: String,
        var isInternal: Boolean = false,
        var isPlugin: Boolean = false,

        var libraryName: String,
        var author: String = "",
        var authorWebsite: String = "",
        var libraryDescription: String = "",
        var libraryVersion: String = "",
        var libraryArtifactId: String = "",
        var libraryWebsite: String = "",
        var license: License? = null,

        var isOpenSource: Boolean = true,
        var repositoryLink: String = "",

        var classPath: String = ""
) : Comparable<Library> {

    override fun compareTo(other: Library): Int {
        return libraryName.compareTo(other.libraryName, ignoreCase = true)
    }

    private fun ifNotEmpty(receiver: String): String? {
        return if (receiver.isEmpty()) null else receiver
    }

    /**
     * transfers the information from one to the other, if set
     */
    fun enchantBy(enchantWith: Library) {
        libraryName = ifNotEmpty(enchantWith.libraryName) ?: libraryName
        author = ifNotEmpty(enchantWith.author) ?: author
        authorWebsite = ifNotEmpty(enchantWith.authorWebsite) ?: authorWebsite
        libraryDescription = ifNotEmpty(enchantWith.libraryDescription) ?: libraryDescription
        libraryVersion = ifNotEmpty(enchantWith.libraryVersion) ?: libraryVersion
        libraryArtifactId = ifNotEmpty(enchantWith.libraryArtifactId) ?: libraryArtifactId
        libraryWebsite = ifNotEmpty(enchantWith.libraryWebsite) ?: libraryWebsite
        license = enchantWith.license ?: license
        isOpenSource = enchantWith.isOpenSource
        repositoryLink = ifNotEmpty(enchantWith.repositoryLink) ?: repositoryLink
    }
}
